src/models/Observation.js

```diff
@@ -335,6 +348,89 @@ class Observation extends Realm.Object {
     } );
   }
 
+  static markRecordUploaded = async ( recordUUID, type, response, realm ) => {
+    const { id } = response.results[0];
+    try {
+      const record = realm.objectForPrimaryKey( type, recordUUID );
+      realm?.write( ( ) => {
+        record.id = id;
+        record._synced_at = new Date( );
+      } );
+    } catch ( e ) {
+      console.log( e, `couldn't mark ${type} uploaded in realm` );
+    }
+  };
+
+  static uploadToServer = async (
+    evidenceUUID: string,
+    type: string,
+    params: Object,
+    apiEndpoint: Function,
+    realm: any
+  ) => {
+    const apiToken = await getJWTToken( false );
+    const options = { api_token: apiToken };
+
+    let response;
+    try {
+      response = await apiEndpoint.create( params, options );
+      await Observation.markRecordUploaded( evidenceUUID, type, response, realm );
+    } catch ( e ) {
+      return JSON.stringify( e.response );
+    }
+    return response;
+  };
+
+  static uploadEvidence = (
+    evidence: Array<Object>,
+    type: string,
+    apiSchemaMapper: Function,
+    observationId: number,
+    apiEndpoint: Function,
+    realm: any
+  ): any => {
+    let response;
+    if ( evidence.length === 0 ) { return; }
+    for ( let i = 0; i < evidence.length; i += 1 ) {
+      const currentEvidence = evidence[i];
+      const evidenceUUID = currentEvidence.uuid;
+      const params = apiSchemaMapper( observationId, currentEvidence );
+      response = Observation.uploadToServer( evidenceUUID, type, params, apiEndpoint, realm );
+    }
+    // eslint-disable-next-line consistent-return
+    return response;
+  };
+
+  static uploadObservation = async obs => {
+    const obsToUpload = Observation.mapObservationForUpload( obs );
+    const apiToken = await getJWTToken( false );
+    const options = { api_token: apiToken };
+
+    // Remove all null values, b/c the API doesn't seem to like them for some
+    // reason (might be an error with the API as of 20220801)
+    const newObs = {};
+    Object.keys( obsToUpload ).forEach( k => {
+      if ( obsToUpload[k] !== null ) {
+        newObs[k] = obsToUpload[k];
+      }
+    } );
+
+    const uploadParams = {
+      observation: { ...newObs },
+      fields: { id: true }
+    };
+
+    let response;
+    try {
+      response = await inatjs.observations.create( uploadParams, options );
+    } catch ( uploadError ) {
+      const body = JSON.parse( await uploadError.response.text( ) );
+      console.error( "[ERROR] Failed to upload observation: ", JSON.stringify( body ) );
+      return body;
+    }
+    return response;
+  }
+
```

Comment:
This doesn't need to block, but in general I don't think network operations belong in the model. A model is a representation of data, so at a minimum it's a class with attributes corresponding to that kind of data, with methods for manipulating that data, and for persisting that data. So CRUD operations belong in the model, and maybe mapping model attributes into different schemas, but probably not things like performing network operations, and definitely not anything related to presentation logic (e.g. a model method would never return a component).

That said, I'm still not sure where this kind of stuff belongs. My gut says a hook, but maybe they belong in some kind of Uploader class that a hook or a component holds onto in state. In Rails this kind of thing would happen in a controller or maybe in some kind of utility class, and in Android this might be encapsulated in a repository of some kind.

Reply:
Ok, good feedback. Maybe we can take a closer look at where to put network operations while working on the wrapper API issue (#151). Hooks or shared hooks make sense to me.